File: utilities/shell/direnv-config/sdk/php/tests/ContractTest.php
<?php

declare(strict_types=1);

namespace Noizu\DirenvConfig\Tests;

use Noizu\DirenvConfig\Backend\NativeBackend;
use Noizu\DirenvConfig\Store;
use Noizu\DirenvConfig\Version;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Yaml\Yaml;

final class ContractTest extends TestCase
{
    private static array $expectations;
    private static string $fixturesDir;

    public static function setUpBeforeClass(): void
    {
        self::$fixturesDir = dirname(__DIR__, 2) . '/contract-tests/fixtures';
        $path = dirname(__DIR__, 2) . '/contract-tests/expectations.yaml';
        self::$expectations = Yaml::parseFile($path);
    }

    /** @return iterable<string, array{array}> */
    public static function contractCases(): iterable
    {
        $fixturesDir = dirname(__DIR__, 2) . '/contract-tests/fixtures';
        $path = dirname(__DIR__, 2) . '/contract-tests/expectations.yaml';
        $data = Yaml::parseFile($path);

        foreach ($data['tests'] as $test) {
            yield $test['name'] => [$test];
        }
    }

    #[\PHPUnit\Framework\Attributes\DataProvider('contractCases')]
    public function testContractExpectation(array $test): void
    {
        $name = $test['name'];

        if (isset($test['expected_version'])) {
            $this->runVersionTest($test);
            return;
        }

        if (isset($test['expected_configs'])) {
            $this->runListConfigsTest($test);
            return;
        }

        if (isset($test['input_path'])) {
            $this->runPathHashTest($test);
            return;
        }

        $this->runGetTest($test);
    }

    private function runGetTest(array $test): void
    {
        $storePath = self::$fixturesDir . '/' . $test['store'];
        $backend = new NativeBackend($storePath);
        $path = $test['path'] ?? null;
        if ($path === 'null') {
            $path = null;
        }

        $result = $backend->get($test['config'], $path);
        $type = $test['type'];

        match ($type) {
            'string' => $this->assertSame($test['expected'], $result),
            'integer' => $this->assertSame((int) $test['expected'], $result),
            'boolean' => $this->assertSame((bool) $test['expected'], $result),
            'null' => $this->assertNull($result),
            'string_array' => $this->assertSame($test['expected'], $result),
            'integer_array' => $this->assertSame(
                array_map(fn($v) => (int) $v, $test['expected']),
                $result
            ),
            'map' => $this->assertExpectedKeys($test, $result),
            default => $this->fail("Unknown type: {$type}"),
        };
    }

    private function assertExpectedKeys(array $test, mixed $result): void
    {
        $this->assertIsArray($result);
        foreach ($test['expected_keys'] as $key) {
            $this->assertArrayHasKey($key, $result, "Missing key: {$key}");
        }
    }

    private function runVersionTest(array $test): void
    {
        $storePath = self::$fixturesDir . '/' . $test['store'];
        $version = Version::read($storePath);
        $this->assertSame((int) $test['expected_version'], $version);
    }

    private function runListConfigsTest(array $test): void
    {
        $storePath = self::$fixturesDir . '/' . $test['store'];
        $backend = new NativeBackend($storePath);
        $configs = $backend->listConfigs();
        sort($configs);
        $expected = $test['expected_configs'];
        sort($expected);
        $this->assertSame($expected, $configs);
    }

    private function runPathHashTest(array $test): void
    {
        $hash = Store::pathToHash($test['input_path']);
        $this->assertSame($test['expected_hash'], $hash);
    }
}

File: sdk/elixir/test/contract_test.exs
defmodule DirenvConfig.ContractTest do
  use ExUnit.Case, async: true

  @fixtures_dir Path.expand("../../contract-tests/fixtures", __DIR__)
  @expectations_file Path.expand("../../contract-tests/expectations.yaml", __DIR__)

  setup_all do
    {:ok, expectations} = YamlElixir.read_from_file(@expectations_file)
    %{tests: expectations["tests"]}
  end

  describe "contract expectations" do
    setup %{tests: tests} do
      %{tests: tests}
    end

    test "all value lookup cases pass", %{tests: tests} do
      tests
      |> Enum.filter(&Map.has_key?(&1, "config"))
      |> Enum.filter(&Map.has_key?(&1, "expected"))
      |> Enum.each(fn tc ->
        store = Path.join(@fixtures_dir, tc["store"])
        config = tc["config"]
        path = tc["path"]

        result = DirenvConfig.Native.get(store, config, path)

        case tc["type"] do
          "null" ->
            assert result == :error,
                   "#{tc["name"]}: expected :error, got #{inspect(result)}"

          "string" ->
            assert {:ok, value} = result, "#{tc["name"]}: expected {:ok, _}, got :error"
            assert to_string(value) == to_string(tc["expected"]),
                   "#{tc["name"]}: expected #{inspect(tc["expected"])}, got #{inspect(value)}"

          "integer" ->
            assert {:ok, value} = result, "#{tc["name"]}: expected {:ok, _}, got :error"
            assert value == tc["expected"],
                   "#{tc["name"]}: expected #{tc["expected"]}, got #{inspect(value)}"

          "boolean" ->
            assert {:ok, value} = result, "#{tc["name"]}: expected {:ok, _}, got :error"
            assert value == tc["expected"],
                   "#{tc["name"]}: expected #{tc["expected"]}, got #{inspect(value)}"

          "string_array" ->
            assert {:ok, value} = result, "#{tc["name"]}: expected {:ok, _}, got :error"
            assert Enum.map(value, &to_string/1) == Enum.map(tc["expected"], &to_string/1),
                   "#{tc["name"]}: expected #{inspect(tc["expected"])}, got #{inspect(value)}"

          "integer_array" ->
            assert {:ok, value} = result, "#{tc["name"]}: expected {:ok, _}, got :error"
            assert value == tc["expected"],
                   "#{tc["name"]}: expected #{inspect(tc["expected"])}, got #{inspect(value)}"

          "map" ->
            assert {:ok, _value} = result, "#{tc["name"]}: expected {:ok, _}, got :error"
        end
      end)
    end

    test "map key expectations", %{tests: tests} do
      tests
      |> Enum.filter(&Map.has_key?(&1, "expected_keys"))
      |> Enum.each(fn tc ->
        store = Path.join(@fixtures_dir, tc["store"])
        config = tc["config"]
        path = tc["path"]

        assert {:ok, value} =
                 DirenvConfig.Native.get(store, config, path),
                 "#{tc["name"]}: expected {:ok, _}, got :error"

        assert is_map(value), "#{tc["name"]}: expected map, got #{inspect(value)}"

        Enum.each(tc["expected_keys"], fn key ->
          assert Map.has_key?(value, key),
                 "#{tc["name"]}: missing key #{inspect(key)} in #{inspect(Map.keys(value))}"
        end)
      end)
    end

    test "version expectations", %{tests: tests} do
      tests
      |> Enum.filter(&Map.has_key?(&1, "expected_version"))
      |> Enum.each(fn tc ->
        store = Path.join(@fixtures_dir, tc["store"])
        version = DirenvConfig.Version.read(store)

        assert version == tc["expected_version"],
               "#{tc["name"]}: expected version #{tc["expected_version"]}, got #{version}"
      end)
    end

    test "config listing expectations", %{tests: tests} do
      tests
      |> Enum.filter(&Map.has_key?(&1, "expected_configs"))
      |> Enum.each(fn tc ->
        store = Path.join(@fixtures_dir, tc["store"])
        assert {:ok, configs} = DirenvConfig.Native.list_configs(store)

        assert configs == tc["expected_configs"],
               "#{tc["name"]}: expected #{inspect(tc["expected_configs"])}, got #{inspect(configs)}"
      end)
    end

    test "path hash expectations", %{tests: tests} do
      tests
      |> Enum.filter(&Map.has_key?(&1, "input_path"))
      |> Enum.each(fn tc ->
        hash = DirenvConfig.Store.path_to_hash(tc["input_path"])

        assert hash == tc["expected_hash"],
               "#{tc["name"]}: expected #{inspect(tc["expected_hash"])}, got #{inspect(hash)}"
      end)
    end
  end
end
